[draft] Consume control node capacity

kdelee opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Consume capacity on control nodes and consider their capacity when scheduling jobs."
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
This PR both consumes capacity on control nodes as well as considers their capacity when scheduling jobs.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
replaces https://github.com/ansible/awx/pull/11620
addresses https://github.com/ansible/awx/issues/10694

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 19.5.2.dev72+gf7b073ba5d
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

kdelee committed:
Consume control node capacity

This is a rough start, I've only tested with container groups
Needs several refactors, but it generally results in desired behavior
of only starting container group jobs when we have a control node with
capacity to run them. May not be too difficult to extend to other job
types that have a seperate control node.

rebeccahhh committed:
change the query for available capacity

kdelee committed:
Add staticmethod to find best control instance

Maybe we can use this elsewhere next

rebeccahhh committed:
allow project & inventory updates to run on control nodes, also edit consume capacity debug message to see what each job gets in terms of capacity

rebeccahhh committed:
fix django query so it excludes successful jobs, because we don't care about them

kdelee committed:
use hostname for control node, not whole object

kdelee committed:
WELL...lets try and keep track ourselves

PS this is not valid python please fix

rebeccahhh committed:
fixed some python syntax, but found some logical errors and started working through those (or maybe my brain is just struggling)

kdelee committed:
Cut out extra loop over instances

We already loop over instances a few times, this cuts out one more and
builds the map of instance:remaining_capacity in an existing loop.

Delete a old key instances_remaining_capacity in the graph, because it
is not being used anywhere. I think this indicates someone in the past
had a similar intention as we have here for keeping track of the
remaining capacity on individual nodes, not just the group.

kdelee committed:
not all Jobs have the controller_node attribute

kdelee committed:
don't reap WF jobs when looking for orphans

workflow jobs don't have a control or execution node, but
they are not "orphaned" in the sense we are worried about in this
clause.

rebeccahhh committed:
adding a warning for when an instance group is not found for a control node, also modified if/else to be more specific around node_type

rebeccahhh committed:
when settings.IS_K8s is true this code will never be reached, rendering it useless.

rebeccahhh committed:
add a comment, hopefully interactive rebase this message out later

kdelee committed:
clean up inefficient debugging in capacity method

kdelee committed:
Refactor two identical loops to one

We were looping over the instances and excluding any disabled ones in
two loops, one after another. Just do the work in the one loop.

kdelee committed:
hype up this function

Not yet used anywhere, but its cooL!

kdelee committed:
use complete words as variable names

kdelee committed:
fix typo causing failures

kdelee committed:
merge three nearly identica data structures

instances_by_hostname
instances_partial
self.real_instances

Were all...mappings of instance host names to some form or another of
data about the instance. Instead of having three of these, lets just get
by with one.

kdelee committed:
Give map of instances a more dignified name

Move the map of instances in the task manager into its own class. Make
the init of the graph using that data a method of that class. Move these
classes that help encapsulate the capacity and ability of all the instance the task
manager will be assigning work to into another file. It would be great
to add some tests of these classes themselves now that they are discreet
units.

do control node assignment in task manager instances class

TaskManagerInstances allready has the data needed for this, so cut out
excessive passing around of this information by doing the control
node assignment in this class.

For the two things called "graph", use two distinctive names

One is a graph of instance groups and the tracking data for what
capacity they have and what we've commited to in the run of the task
manager. The other is a DependencyGraph for tracking job dependencies.

Also, use the same method to zero out the graph in the two places we
might want to do it. The main reason we would ever actually zero out the
graph in awx/main/managers.py is in tests where we have not already
initialized the instance group capacity graph.

kdelee committed:
Consume capacity for control nodes for all jobs

Choose control nodes before we call start_task for all job types.
Also track how much capacity we are consuming for those jobs.

In the original issue Alan mentions not counting the full task impact on
the controlnode, so I do that here for non-container group jobs.

Still counting the full task impact on container group jobs, but that
can change.

AlanCoding commented on awx/main/models/ha.py:
```diff
@@ -145,7 +146,12 @@ def get_absolute_url(self, request=None):
-        return sum(x.task_impact for x in UnifiedJob.objects.filter(execution_node=self.hostname, status__in=('running', 'waiting')))
+        capacity_consumed = sum(
+            x.task_impact
+            for x in UnifiedJob.objects.filter((Q(controller_node=self.hostname) | Q(execution_node=self.hostname)) & Q(status__in=('running', 'waiting')))
```
Do you intend the filter here to return duplicates? Because if a job is on a hybrid node, it _should_ be double-counted. It should be counted by `task_impact` for the execution part, and the constant value for the control part, so at first glance I don't see how this is going to work.

What is the intended capacity consumption for the control nodes? I'd like to start with some docs about that.

kdelee replied:
Good call on double counting for hybrid nodes. I think I need to work on that, I originally was just working on container group jobs where that is never the case.

I had started with just counting the task impact as the same on the control node as the execution node, because that was simpler.

Problem with going with same control node task impact as execution:
  - we'll be able to run less jobs. But maybe for good reason, if you really are running a job against 1,000 hosts, the events coming back from that job are very likely MUCH more than a job running against 1 host. So scaling the control node impact w/ `task.task_impact` seems good.

Problem with going with small, constant impact for control nodes:
  - we'll not put much of a throttle on controlling jobs
  - doesn't scale the task impact in same way as we scale it for execution node capacity impact
  - in k8s, the whole way we are calculating capacity of control nodes is wrong https://github.com/ansible/awx/issues/11640 so a tiny constant impact won't show much improvement re; not overloading conttrol plane jobs

kdelee replied:
I have addressed this on hybrid nodes in https://github.com/ansible/awx/pull/11651/commits/1fe442270db44279fe3c64cbcdcbd1e20b959709

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -36,9 +35,8 @@
 from awx.main.utils import get_type_for_model, task_manager_bulk_reschedule, schedule_task_manager
 from awx.main.utils.common import create_partition
 from awx.main.signals import disable_activity_stream
-from awx.main.scheduler.dependency_graph import DependencyGraph
```
Wouldn't this leave this class orphaned? No uses would remain. If you do this, you should delete that class (probably the whole module), but I might have more questions before jumping into that.

kdelee replied:
The class does get used in the new file `awx/main/scheduler/task_manager_instances.py`

rebeccahhh committed:
add test to verify that when there are 2 jobs and only capacity for one that one will move into waiting and the other stays in pending

kdelee committed:
Treat hybrid nodes correctly

Hybrid nodes need the capacity impact of both the control and execution
tasks.

Also, use a constant task impact for control nodes as was originally
proposed in the issue https://github.com/ansible/awx/issues/10694

kdelee committed:
use control node task impact setting everywhere

This is to be consistent with the decision to have control node tasks
have a constant impact, as opposed to the same impact as
task.task_impact which is the cost of running the job itself.

AlanCoding commented on awx/main/models/ha.py:
```diff
@@ -145,7 +146,19 @@ def get_absolute_url(self, request=None):
+                for x in UnifiedJob.objects.filter(Q(controller_node=self.hostname) & Q(status__in=('running', 'waiting')))
+            )
+        logger.debug(f"{execution_capacity_consumed} execution capacity and {control_capacity_consumed} control capacity consumed on {self.hostname}")
+        return execution_capacity_consumed + control_capacity_consumed
```
Okay, I think this looks valid now. I need to think about performance, but this at least seems correct.

kdelee committed:
Move instance capacity graph to its own class

Now the in-flight capacity of nodes and instance groups is handled
all in TaskManagerInstances and InstanceGroupCapacityGraph

The selection of execution nodes is still being done in the old way,
I just moved the code to the TaskManagerInstances class.

This commit does not introduce any behavior change except we consume
capacity on the execution node instance group just before we enter into
"start_task" instead of at the end, but this seems to be desired
as 'waiting' tasks should also consume capacity.

kdelee committed:
no longer choosing random control node

Neither of these functions are used anywhere. We are choosing
the control node based off of most available capacity as tracked in
TaskManagerInstances.

kdelee committed:
Move static methods to TaskManagerInstances class

Finally understood that we are tracking the in-flight resource
consumption of instances via SimpleNamespace objects we are carrying
around in ig_capacity_graph[ig_name][instances]. Moved the static
methods that choose what the best one of these is into the
TaskManagerInstances class so all the logic that determines what
instance we choose and what the in-flight capacity consumption is all
under one roof.

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,213 @@
+        for rampart_group in InstanceGroup.objects.prefetch_related('instances'):
+            InstanceGroupManager.zero_out_group(self.ig_capacity_graph, rampart_group.name, breakdown)
+            # Didn't move the init of DependencyGraph to InstanceGroupManager because of circular import
+            self.ig_capacity_graph[rampart_group.name]['dependency_graph'] = DependencyGraph()
```
I'm trying to figure out, what is a dependency graph, and why does an instance group need to have one attached?

I can see why there is a global dependency graph for jobs in general. This dictates blocking logic, which I see breadcrumbs of. But why put it here? What connection does dependency / blocking logic have to instance groups?

rebeccahhh replied:
This replaces the graphing on line 106 of `awx/main/scheduler/task_manager.py` (the call to this is on line 60).
https://github.com/ansible/awx/blob/devel/awx/main/scheduler/dependency_graph.py <- file the function comes from
All that to say, this paradigm existed before this refactor.

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,213 @@
+            InstanceGroupManager.zero_out_group(self.ig_capacity_graph, rampart_group.name, breakdown)
+            # Didn't move the init of DependencyGraph to InstanceGroupManager because of circular import
+            self.ig_capacity_graph[rampart_group.name]['dependency_graph'] = DependencyGraph()
+            for instance in rampart_group.instances.filter(enabled=True).order_by('hostname'):
```
I'm never entirely clear on behavior with experimenting, but if you do `rampart_group.instances.filter(enabled=True).order_by('hostname')`, do you hit the cache populated by `InstanceGroup.objects.prefetch_related('instances')`?

With the complexity here... I don't think so. If you want enabled instances ordered by hostname, then you need to do that in python. The thing I know you can safely reference without performance penalties is `rampart_group.instances.all()`

kdelee replied:
I really don't think the order matters, this is just what the code was previously doing. I'll experiment w/o the order_by

AlanCoding commented:
I feel like this is getting to be kind of nuts. I regret that I introduced things like `consumed_control_capacity`, and seeing the lengths gone to here, I want to remove it.

You need to assign a pending job's instances, `control_node` in particular. What I would like to see happen is:
 - from whatever in-memory data structure you have, get the "controlplane" group. This is the group you select the node from.
 - based on available capacity of instances in that group, pick the best node, or do not pick any node

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -511,39 +444,83 @@ def process_pending_tasks(self, pending_tasks):
                 continue
 
             for rampart_group in preferred_instance_groups:
```
Even before this line, I think we can assign `controller_node`. I mean, not even in the loop. We don't need to consider the preferred instance groups, that's just over-complicating the problem. All jobs, project updates, container group jobs, regular jobs, all need a control node. If a control node can't be found, you can't dispatch _any_ jobs. You might as well stop processing pending tasks in general.

kdelee replied:
doing this now

rebeccahhh committed:
add skeleton stuff for new test

rebeccahhh committed:
test works now

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,236 @@
+    def best_node(self):
+        heap = list(self.nodes)
+        heapq.heapify(heap)
+        return heapq.nlargest(1, heap).pop()
```
I would really prefer to keep the paradigms we have of

1. try to find a node where the job "fits"
2. find largest idle instance

I don't see why that won't work for control nodes as well. You don't need to reuse the existing logic for this, you could write new methods that do this, but the selection algorithm is something I would prefer to not alter in this refacoring.

rebeccahhh replied:
Is there a technical reason for your preference?

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,236 @@
+            self.instances_partial[instance.hostname] = SimpleNamespace(
+                obj=instance,
+                node_type=instance.node_type,
+                remaining_capacity=instance.remaining_capacity,
```
There are going to be :dragon: s here

https://github.com/ansible/awx/blob/2de5ffc8d94d592498c32b8c3d68a82470d57763/awx/main/models/ha.py#L146-L152

You're using a property on the Instance model which is _very_ inefficient. It queries the list of jobs anew, which is really really counter to the purpose of the task manager. The task manager has complex logic to load instances, instance groups, and jobs. It is _most_ important that we don't re-load jobs multiple times. You can't just create a new data structure for instance capacity because the hard part is integrating it with the list of jobs that the task manager fetches.

shanemcd commented:
I am growing a bit concerned with the scope of this PR. It's not clear to me which parts are necessary, and what is just refactoring. 

This is a very critical part of the application. Any new patterns or large refactoring should be done in isolation so that we can more easily observe the impact on performance.

I would like to take a step back and reassess the problem we are trying to solve and see if there is a more direct path that does not involve such a large diff.

kdelee committed:
Select control node before look for execution

If its a project update, don't  loop over instance groups, go ahead and
start task.
If we find a hybrid execution node, prefer that as the control node.

kdelee committed:
ugly but works

will refactor to not use loop later

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -509,50 +442,91 @@ def process_pending_tasks(self, pending_tasks):
+                # TODO: if we break here, we don't get to do
+                # task.log_lifecycle("needs_capacity")
+                # for all the tasks
+                break
```
I know I said you should `break` here, but you can't because it would halt processing of workflow jobs. Best to `continue` instead.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -509,50 +442,91 @@ def process_pending_tasks(self, pending_tasks):
+                break
+
+            if isinstance(task, ProjectUpdate):
+                controller_node = self.task_manager_instances.assign_task_to_control_node(task, self.ig_capacity_graph, controller_node)
```
You already picked this, I don't see any need to pick it again. Can't you just delete this line?

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,252 @@
+            if i.node_type not in (task.capacity_type, 'hybrid'):
+                continue
+            if i.remaining_capacity >= task.task_impact and (
+                instance_most_capacity is None or i.remaining_capacity > instance_most_capacity.remaining_capacity
```
Instead of considering `i.remaining_capacity >= task.task_impact` as the criteria here, I think you should consider the criteria to be `i.remaining_capacity >= task.task_impact + settings.AWX_CONTROL_NODE_TASK_IMPACT` for hybrid nodes.

This should be actually pretty easy to do, although you'll be adding a little bit more python code to this method. This avoids any situation where we "fit" something but it won't actually fit because the control capacity consumption pushes it over.

This is a key change that I alluded to but probably didn't communicate effectively. If it doesn't make sense, let me know. It's really important for what needs to happen here.

AlanCoding commented on awx/main/scheduler/task_manager.py:
```diff
@@ -597,27 +571,23 @@ def reap_jobs_from_orphaned_instances(self):
+                and j.status in ['waiting', 'running']
+            ):
+                logger.error(f'Job {j.id} has no control node, but is in status {j.status}. The control node pod must have been lost; reaping {j.log_format}')
+                reap_job(j, 'error')
```
Reaping a job at this location seems very suspect to me. How could the control node have gone away? You just ran in a loop to select it. This doesn't add up to me at all.

kdelee committed:
Track capacity of instances not instance groups

This more forcefully insists that there must be a controlplane instance
group. The tests to not ensure this. I've updated a few tests, and as I
do ensure that there is an instance group named controlplane and there
is some hybrid or control instance in it, they are passing.

So this is incomplete in that more tests need to be updated and may
reveal some corner cases to fix, but in
general in my local dev environment jobs/project updates etc are
running and consuming capacity.

AlanCoding commented on awx/main/scheduler/task_manager_instances.py:
```diff
@@ -0,0 +1,204 @@
+            return largest_instance.hostname
+        return None
+
+    def fit_task_to_most_remaining_capacity_instance(self, task, instance_group, capacity_type, impact=None):
```
After tinkering with this a bit today, I have to say that the _old_ `fit_task_to_most_remaining_capacity_instance` method was _beautifully_ isolated and easy to read. It took in 2 arguments - a job, and a list of instances. It was a _static_ method, meaning that it didn't require initializing a class or anything. It got the `capacity_type` from the task itself.

Was this method _in the right place_? No. Attaching it to the manager was confusing. IMO, a better place would be a new module with shared methods, like `awx/main/scheduler/fitting.py`, where it has nothing other than shared static methods.

I'm pro-refactoring, but I'm reading down into this method and seeing new loop structures, and it doesn't seem to be getting any more simple. A list of instances to consider is simple.

kdelee commented:
Closing this since the refactor is overshadowing the main point, which is to select the control node earlier and based on capacity and track consumed capacity for control nodes.

Replacement: https://github.com/ansible/awx/pull/11665

AlanCoding commented:
Our main issue for task manager refactoring is https://github.com/ansible/awx/issues/11306

I think the sub-components of that issue are still wildly unfocused but we have discussed several concrete actions that we _should_ and _need_ to take here. As I follow up from this, my goal is to address those quality points @kdelee discovered through our process so that we don't lose the important things in a sea of non-specific issues or giant code diffs.
